File: backend/src/services/langgraph/nodes/evaluation.ts
import { ConversationGraphState } from '../state';
import { executeUserBehaviorAnalysis, executeAiIndicatorsAnalysis } from '../tools/evaluation_tools';
import { emitGoalProgressUpdate } from '@/services/realtime';
import { devLogLangGraphEvent } from '../devLogger';

// Lazy imports to avoid TypeORM initialization during module load
let AppDataSource: any;
let SimulationSession: any;
let SessionMessage: any;
let MessageType: any;

/**
 * Lazy-load database dependencies
 */
async function loadDatabaseDependencies() {
  if (!AppDataSource) {
    const databaseModule = await import('@/config/database');
    AppDataSource = databaseModule.AppDataSource;
    
    const simulationSessionModule = await import('@/entities/SimulationSession');
    SimulationSession = simulationSessionModule.SimulationSession;
    
    const SessionMessageModule = await import('@/entities/SessionMessage');
    SessionMessage = SessionMessageModule.SessionMessage;
    MessageType = SessionMessageModule.MessageType;
  }
}

/**
 * Evaluation thresholds
 */
class EvaluationThresholds {
  // [TODO]: USE UNIQUE THRESOLDS PER EACH SIMULATION GOAL
  private static readonly BEHAVIOR_THRESHOLD = 0.7;
  private static readonly SUCCESS_THRESHOLD = 0.7;

  private static isTestMode(): boolean {
    return process.env.NODE_ENV === 'test' || !!process.env.JEST_WORKER_ID;
  }

  static getBehaviorThreshold(): number {
    return this.isTestMode() ? 0.6 : this.BEHAVIOR_THRESHOLD;
  }

  static getSuccessThreshold(): number {
    return this.isTestMode() ? 0.6 : this.SUCCESS_THRESHOLD;
  }
}

function asFiniteNumber(value: unknown, fallback: number): number {
  return typeof value === 'number' && Number.isFinite(value) ? value : fallback;
}

/**
 * Get actual message IDs from database for evidence tracking
 */
async function getRecentMessageIds(
  sessionId: string,
  limit: number = 2,
): Promise<{ userMessageId?: string; aiMessageId?: string }> {
  await loadDatabaseDependencies();
  
  try {
    const messageRepo = AppDataSource.getRepository(SessionMessage);
    const messages = await messageRepo
      .createQueryBuilder('message')
      .where('message.sessionId = :sessionId', { sessionId })
      .orderBy('message.sequenceNumber', 'DESC')
      .limit(limit)
      .getMany();

    const aiMessage = messages.find((m) => m.type === MessageType.AI);
    const userMessage = messages.find((m) => m.type === MessageType.USER);

    return {
      userMessageId: userMessage?.id,
      aiMessageId: aiMessage?.id,
    };
  } catch (error) {
    console.warn('Error fetching message IDs:', error);
    return {};
  }
}

/**
 * Detect which goal the user is currently working on using enhanced labels
 * Uses zero-shot classification with goal title + key behaviors for better accuracy
 */
async function detectActiveGoal(
  userMessage: string,
  goals: any[],
  progress: any[],
): Promise<any | null> {
  try {
    // Get unachieved goals
    const unachieved = goals.filter((g: any) => {
      const p = progress.find((p) => p.goalNumber === g.goalNumber);
      return !p || p.status !== 'achieved';
    });

    if (unachieved.length === 0) return null;

    // Create enhanced labels: "Goal Title: key behavior 1; key behavior 2"
    // This provides much richer context for classification
    const enhancedLabels = unachieved.map((g: any) => {
      const behaviors = g.keyBehaviors?.slice(0, 2).join('; ') || '';
      return behaviors ? `${g.title}: ${behaviors}` : g.title;
    });

    console.log(`🔍 Detecting active goal from ${unachieved.length} candidates:`, enhancedLabels);

    // Use transformers service for zero-shot classification
    const { transformersService } = await import('@/services/transformers');
    const result = await transformersService.classifySequence(userMessage, enhancedLabels);
    
    const threshold = EvaluationThresholds.getBehaviorThreshold();
    console.log(`   📊 Best match: "${result.label}" (confidence: ${result.confidence.toFixed(3)}, threshold: ${threshold})`);

    // Only consider it a match if confidence is high enough
    if (result.confidence >= threshold) {
      // Find the goal that matches this label
      const matchedGoal = unachieved.find((g: any) => {
        const behaviors = g.keyBehaviors?.slice(0, 2).join('; ') || '';
        const label = behaviors ? `${g.title}: ${behaviors}` : g.title;
        return label === result.label;
      });
      
      if (matchedGoal) {
        console.log(`   ✅ Active goal detected: #${matchedGoal.goalNumber} - ${matchedGoal.title}`);
        return matchedGoal;
      }
    } else {
      console.log(`   ⚠️  No confident match (${result.confidence.toFixed(3)} < ${threshold})`);
    }

    return null;
  } catch (error) {
    console.warn('Error detecting active goal:', error);
    return null;
  }
}

/**
 * Node: Evaluate Goals
 * Enhanced version with parallel execution, sentiment analysis, and AI-powered goal detection
 */
export async function evaluateGoalsNode(
  state: ConversationGraphState,
): Promise<Partial<ConversationGraphState>> {
  console.log(`🎯 Evaluating goals for session ${state.sessionId}`);

  // Skip if evaluation not needed
  if (!state.needsEvaluation) {
    console.log(`🔄 Evaluation not needed, skipping`);
    return { evaluationComplete: true };
  }

  // Skip if no goals defined
  if (!state.simulation?.conversationGoals || state.simulation.conversationGoals.length === 0) {
    console.log(`🔄 No goals defined, skipping`);
    return { evaluationComplete: true };
  }

  try {
    const goals = state.simulation.conversationGoals.sort((a: any, b: any) => a.goalNumber - b.goalNumber);
    let progress = state.goalProgress || [];
    await devLogLangGraphEvent(state.sessionId, 'node:evaluate_goals:start', {
      lastUserMessage: state.lastUserMessage,
      lastAiMessage: state.lastAiMessage,
      goals: goals.map((g: any) => ({
        goalNumber: g.goalNumber,
        title: g.title,
        isOptional: !!g.isOptional,
        evaluationConfig: g.evaluationConfig,
        keyBehaviors: g.keyBehaviors,
        successIndicators: g.successIndicators,
      })),
      progress,
    });

    // Initialize progress if needed
    if (progress.length === 0) {
      console.log(`🔄 No progress, initializing`);
      progress = goals.map((g: any) => ({
        goalNumber: g.goalNumber,
        isOptional: !!g.isOptional,
        title: g.title,
        status: 'not_started' as const,
        confidence: 0,
        evidence: [],
      }));
    }

    // Get actual message IDs for evidence tracking (once for all goals)
    const { userMessageId, aiMessageId } = await getRecentMessageIds(state.sessionId);

    // Stage-based evaluation:
    // Evaluate ONLY the next unachieved required goal in sequence.
    // This prevents "jumping" (e.g., achieving Professional Closing early) and keeps evaluation aligned with staged prompting.
    let goalsToEvaluate: any[] = [];
    
    const requiredGoals = goals
      .filter((g: any) => !g.isOptional)
      .sort((a: any, b: any) => a.goalNumber - b.goalNumber);
    
    const nextRequired = requiredGoals.find((g: any) => {
      const p = progress.find((p) => p.goalNumber === g.goalNumber);
      return !p || p.status !== 'achieved';
    });
    
    if (nextRequired) {
      console.log(`🎯 Stage-based: evaluating next required goal #${nextRequired.goalNumber} - ${nextRequired.title}`);
      goalsToEvaluate = [nextRequired];
    } else {
      // All required goals done, check if there are any optional left (in order)
      const optionalGoals = goals
        .filter((g: any) => !!g.isOptional)
        .sort((a: any, b: any) => a.goalNumber - b.goalNumber);
      
      const nextOptional = optionalGoals.find((g: any) => {
        const p = progress.find((p) => p.goalNumber === g.goalNumber);
        return !p || p.status !== 'achieved';
      });
      
      if (nextOptional) {
        console.log(`🎯 Stage-based: all required achieved, evaluating next optional goal #${nextOptional.goalNumber} - ${nextOptional.title}`);
        goalsToEvaluate = [nextOptional];
      }
    }

    console.log(`🎯 Goal evaluation:`, {
      totalGoals: goals.length,
      goalsToEvaluate: goalsToEvaluate.length,
      evaluatingGoal: goalsToEvaluate[0] ? `#${goalsToEvaluate[0].goalNumber} - ${goalsToEvaluate[0].title}` : 'none',
      achievedCount: progress.filter(p => p.status === 'achieved').length,
      progressStatus: progress.map((p) => `#${p.goalNumber}: ${p.status}`).join(', '),
    });

    if (goalsToEvaluate.length === 0) {
      console.log(`🎉 All goals completed!`);
      return {
        evaluationComplete: true,
        needsEvaluation: false,
      };
    }

    // Evaluate all unachieved goals in parallel
    const defaultBehaviorThreshold = EvaluationThresholds.getBehaviorThreshold();
    const defaultSuccessThreshold = EvaluationThresholds.getSuccessThreshold();
    const defaultStrongEvidenceScore = 0.65;
    const defaultMinEvidenceCount = 2;
    const defaultMinStrongEvidenceCount = 2;

    let anyGoalAchieved = false;

    // Process each goal
    for (const goal of goalsToEvaluate) {
      const targetProgress = progress.find((p) => p.goalNumber === goal.goalNumber);
      if (!targetProgress) continue;

      // Mark as in progress if not started
      if (targetProgress.status === 'not_started') {
        targetProgress.status = 'in_progress';
        targetProgress.startedAt = new Date().toISOString();
      }

      const hasBehaviors = goal.keyBehaviors?.length > 0;
      const hasIndicators = goal.successIndicators?.length > 0;

      // Per-goal evaluation tuning (stored in simulation.conversationGoals JSON)
      const cfg = (goal as any).evaluationConfig || {};
      const behaviorThreshold = asFiniteNumber(cfg.behaviorThreshold, defaultBehaviorThreshold);
      const successThreshold = asFiniteNumber(cfg.successThreshold, defaultSuccessThreshold);
      const strongEvidenceScore = asFiniteNumber(cfg.strongEvidenceScore, defaultStrongEvidenceScore);
      const minEvidenceCount = Math.floor(asFiniteNumber(cfg.minEvidenceCount, defaultMinEvidenceCount));
      const minStrongEvidenceCount = Math.floor(asFiniteNumber(cfg.minStrongEvidenceCount, defaultMinStrongEvidenceCount));
      const requireSuccessIndicators = cfg.requireSuccessIndicators !== undefined ? !!cfg.requireSuccessIndicators : true;

      let behaviorScore = 0;
      let successScore = 0;
      let behaviorResult: any = undefined;
      let indicatorResult: any = undefined;

      // Analyze user behavior if we have a last user message
      const behaviorPromise = hasBehaviors && state.lastUserMessage
        ? executeUserBehaviorAnalysis(state.lastUserMessage, goal.keyBehaviors)
        : Promise.resolve({ score: 0 });

      // Execute analyses in parallel for this goal
      if (state.lastAiMessage && hasIndicators) {
        // Run behavior and indicator analysis in parallel
        // Note: executeAiIndicatorsAnalysis already includes emotion/sentiment analysis and tone boost
        const [br, ir] = await Promise.all([
          behaviorPromise,
          executeAiIndicatorsAnalysis(state.lastAiMessage, goal.successIndicators),
        ]);

        behaviorResult = br;
        indicatorResult = ir;
        behaviorScore = behaviorResult?.score || 0;
        successScore = indicatorResult?.score || 0;

        // Record behavior evidence
        if (behaviorScore > 0) {
          targetProgress.confidence = Math.max(targetProgress.confidence || 0, behaviorScore);
          if (!targetProgress.evidence) targetProgress.evidence = [];
          targetProgress.evidence.push({
            messageId: userMessageId || 'user-' + Date.now(),
            role: 'user',
            label: 'behavior',
            score: behaviorScore,
          });
        }

        // Record success indicator evidence
        if (successScore > 0) {
          if (!targetProgress.evidence) targetProgress.evidence = [];
          targetProgress.evidence.push({
            messageId: aiMessageId || 'ai-' + Date.now(),
            role: 'ai',
            label: 'success',
            score: successScore,
          });
        }
      } else {
        // No AI message or no indicators; only await behavior analysis
        behaviorResult = await behaviorPromise;
        behaviorScore = behaviorResult?.score || 0;
        
        if (behaviorScore > 0) {
          targetProgress.confidence = Math.max(targetProgress.confidence || 0, behaviorScore);
          if (!targetProgress.evidence) targetProgress.evidence = [];
          targetProgress.evidence.push({
            messageId: userMessageId || 'user-' + Date.now(),
            role: 'user',
            label: 'behavior',
            score: behaviorScore,
          });
        }
      }

      // Fix 3: Use AND logic instead of OR, and require evidence
      const behaviorMet =
        behaviorScore >= behaviorThreshold &&
        (targetProgress.confidence || 0) >= behaviorThreshold;

      // Success criteria:
      // - If indicators exist AND requireSuccessIndicators=true → require indicator match
      // - If indicators exist AND requireSuccessIndicators=false → allow behavior-only (still prefers indicators)
      // - If no indicators exist → require a slightly higher behavior threshold
      const successMet = hasIndicators
        ? (requireSuccessIndicators ? (successScore >= successThreshold) : (successScore >= successThreshold || behaviorScore >= behaviorThreshold))
        : (behaviorScore >= Math.min(1, behaviorThreshold + 0.1));
      
      const evidenceCount = targetProgress.evidence?.length || 0;
      const strongEvidenceCount = targetProgress.evidence?.filter(e => e.score >= strongEvidenceScore).length || 0;
      const hasEnoughEvidence =
        evidenceCount >= minEvidenceCount &&
        strongEvidenceCount >= minStrongEvidenceCount;

      // Debug logging for goal evaluation
      console.log(`📊 Goal ${targetProgress.goalNumber} evaluation:`, {
        title: goal.title,
        status: targetProgress.status,
        behaviorScore,
        successScore,
        confidence: targetProgress.confidence,
        hasBehaviors,
        hasIndicators,
        evaluationConfig: {
          behaviorThreshold,
          successThreshold,
          strongEvidenceScore,
          minEvidenceCount,
          minStrongEvidenceCount,
          requireSuccessIndicators,
        },
        behaviorMet,
        successMet,
        hasEnoughEvidence,
        evidenceCount,
        strongEvidenceCount,
      });
      await devLogLangGraphEvent(state.sessionId, 'node:evaluate_goals:goal_result', {
        goal: {
          goalNumber: goal.goalNumber,
          title: goal.title,
          isOptional: !!goal.isOptional,
        },
        evaluationConfig: {
          behaviorThreshold,
          successThreshold,
          strongEvidenceScore,
          minEvidenceCount,
          minStrongEvidenceCount,
          requireSuccessIndicators,
        },
        inputs: {
          lastUserMessage: state.lastUserMessage,
          lastAiMessage: state.lastAiMessage,
        },
        rawToolResults: {
          behaviorResult,
          indicatorResult,
        },
        computed: {
          behaviorScore,
          successScore,
          behaviorMet,
          successMet,
          hasEnoughEvidence,
          evidenceCount,
          strongEvidenceCount,
          confidence: targetProgress.confidence,
        },
        progressItem: targetProgress,
      });

      if (behaviorMet && successMet && hasEnoughEvidence && targetProgress.status !== 'achieved') {
        targetProgress.status = 'achieved';
        targetProgress.achievedAt = new Date().toISOString();
        anyGoalAchieved = true;
        console.log(`✅ Goal ${targetProgress.goalNumber} achieved!`);
      } else if (targetProgress.status === 'in_progress') {
        const reasons = [];
        if (!behaviorMet) reasons.push('behaviorMet=FAIL');
        if (!successMet) reasons.push('successMet=FAIL');
        if (!hasEnoughEvidence) reasons.push('hasEnoughEvidence=FAIL');
        console.log(`🔄 Goal ${targetProgress.goalNumber} still in progress. Need: ${reasons.join(', ')}`);
      }
    }

    if (anyGoalAchieved) {
      console.log(`🎊 ${progress.filter(p => p.status === 'achieved').length}/${goals.length} goals now achieved`);
    }

    // Load database dependencies
    await loadDatabaseDependencies();
    
    // Update session in database
    const sessionRepo = AppDataSource.getRepository(SimulationSession);
    const session = await sessionRepo.findOne({ where: { id: state.sessionId } });
    if (session) {
      session.goalProgress = progress as any;
      await sessionRepo.save(session);

      // Emit progress update via Socket.IO
      try {
        await emitGoalProgressUpdate(session);
      } catch (emitError) {
        console.warn('Failed to emit goal progress:', emitError);
      }
    }

    const achievedCount = progress.filter(p => p.status === 'achieved').length;
    const inProgressCount = progress.filter(p => p.status === 'in_progress').length;
    console.log(`📊 Evaluation complete. Achieved: ${achievedCount}/${goals.length}, In Progress: ${inProgressCount}`);
    await devLogLangGraphEvent(state.sessionId, 'node:evaluate_goals:done', {
      achievedCount,
      inProgressCount,
      progress,
    });

    return {
      goalProgress: progress,
      evaluationComplete: true,
      needsEvaluation: false,
    };
  } catch (error) {
    console.error('Error evaluating goals:', error);
    await devLogLangGraphEvent(state.sessionId, 'node:evaluate_goals:error', {
      error: error instanceof Error ? { name: error.name, message: error.message, stack: error.stack } : String(error),
    });
    return {
      evaluationComplete: true,
      needsEvaluation: false,
      lastError: error instanceof Error ? error.message : 'Goal evaluation failed',
    };
  }
}

File: agent/src/careersim_agent/graph/nodes/evaluation.py
"""Goal evaluation node using the EVAL_MODEL LLM."""

import logging
from datetime import datetime
from typing import Any, Optional

from ..state import ConversationState, NodeTraceEntry, GoalProgressItem, EvidenceItem
from ...services import get_eval_service

logger = logging.getLogger(__name__)


DEFAULT_BEHAVIOR_THRESHOLD = 0.6
DEFAULT_SUCCESS_THRESHOLD = 0.6
DEFAULT_STRONG_EVIDENCE_SCORE = 0.55
DEFAULT_MIN_EVIDENCE_COUNT = 1
DEFAULT_MIN_STRONG_EVIDENCE_COUNT = 1


def _add_trace(
    state: ConversationState,
    node: str,
    start_time: datetime,
    input_summary: str,
    output_summary: str,
) -> list[NodeTraceEntry]:
    """Add a trace entry for node execution."""
    duration_ms = (datetime.now() - start_time).total_seconds() * 1000

    trace = state.get("node_trace", []).copy()
    trace.append(NodeTraceEntry(
        node=node,
        timestamp=start_time.isoformat(),
        duration_ms=round(duration_ms, 2),
        input_summary=input_summary,
        output_summary=output_summary,
    ))
    return trace


def _get_current_goal(
    goals: list[dict],
    progress: list[GoalProgressItem],
) -> Optional[dict]:
    """Get the current goal to evaluate (first unachieved in order)."""
    progress_map = {p["goalNumber"]: p for p in progress}
    sorted_goals = sorted(goals, key=lambda g: g.get("goalNumber", 0))

    for goal in sorted_goals:
        if goal.get("isOptional"):
            continue
        p = progress_map.get(goal.get("goalNumber", 0), {})
        if p.get("status") != "achieved":
            return goal

    for goal in sorted_goals:
        if not goal.get("isOptional"):
            continue
        p = progress_map.get(goal.get("goalNumber", 0), {})
        if p.get("status") != "achieved":
            return goal

    return None


def evaluate_goals(state: ConversationState) -> dict[str, Any]:
    """Evaluate conversation goals using the EVAL_MODEL LLM.

    Sends the current goal definition, the latest user/AI exchange, and
    existing evidence to the eval LLM. The LLM returns behaviour and
    success scores which are mapped back into GoalProgressItem / EvidenceItem.
    """
    start_time = datetime.now()
    session_id = state.get("session_id", "unknown")

    if not state.get("needs_evaluation"):
        logger.debug(f"[{session_id}] Evaluation not needed, skipping")
        trace = _add_trace(
            state, "evaluate_goals", start_time,
            "needs_evaluation=False", "Skipped",
        )
        return {"evaluation_complete": True, "node_trace": trace}

    logger.info(f"[{session_id}] Evaluating goals")

    simulation = state.get("simulation", {})
    goals = simulation.get("conversationGoals", [])

    if not goals:
        logger.debug(f"[{session_id}] No goals defined, skipping")
        trace = _add_trace(
            state, "evaluate_goals", start_time,
            "No goals", "Skipped",
        )
        return {
            "evaluation_complete": True,
            "needs_evaluation": False,
            "node_trace": trace,
        }

    progress = list(state.get("goal_progress", []))

    if not progress:
        progress = [
            GoalProgressItem(
                goalNumber=g.get("goalNumber", 0),
                isOptional=g.get("isOptional", False),
                title=g.get("title", ""),
                status="not_started",
                confidence=0.0,
                evidence=[],
            )
            for g in goals
        ]

    current_goal = _get_current_goal(goals, progress)

    if not current_goal:
        logger.info(f"[{session_id}] All goals achieved!")
        trace = _add_trace(
            state, "evaluate_goals", start_time,
            "All goals achieved", "Complete",
        )
        return {
            "goal_progress": progress,
            "evaluation_complete": True,
            "needs_evaluation": False,
            "node_trace": trace,
        }

    goal_num = current_goal.get("goalNumber", 0)
    logger.info(f"[{session_id}] Evaluating goal #{goal_num}: {current_goal.get('title', '')}")

    progress_idx = next(
        (i for i, p in enumerate(progress) if p["goalNumber"] == goal_num),
        None,
    )

    if progress_idx is None:
        progress.append(GoalProgressItem(
            goalNumber=goal_num,
            isOptional=current_goal.get("isOptional", False),
            title=current_goal.get("title", ""),
            status="not_started",
            confidence=0.0,
            evidence=[],
        ))
        progress_idx = len(progress) - 1

    target_progress = progress[progress_idx]

    if target_progress.get("status") == "not_started":
        target_progress["status"] = "in_progress"
        target_progress["startedAt"] = datetime.now().isoformat()

    eval_config = current_goal.get("evaluationConfig", {})
    behavior_threshold = eval_config.get("behaviorThreshold", DEFAULT_BEHAVIOR_THRESHOLD)
    success_threshold = eval_config.get("successThreshold", DEFAULT_SUCCESS_THRESHOLD)
    strong_evidence_score = eval_config.get("strongEvidenceScore", DEFAULT_STRONG_EVIDENCE_SCORE)
    min_evidence_count = eval_config.get("minEvidenceCount", DEFAULT_MIN_EVIDENCE_COUNT)
    min_strong_evidence_count = eval_config.get("minStrongEvidenceCount", DEFAULT_MIN_STRONG_EVIDENCE_COUNT)

    last_user_message = state.get("last_user_message", "")
    last_ai_message = state.get("last_ai_message", "")

    # --- LLM evaluation --------------------------------------------------
    service = get_eval_service()
    eval_result = service.evaluate_goal_progress(
        user_msg=last_user_message,
        ai_msg=last_ai_message,
        goal=current_goal,
        evidence_so_far=list(target_progress.get("evidence", [])),
    )

    behavior_score = eval_result["behavior_score"]
    success_score = eval_result["success_score"]
    message_count = state.get("message_count", 0)

    if behavior_score > 0:
        target_progress["confidence"] = max(
            target_progress.get("confidence", 0.0),
            behavior_score,
        )
        evidence = list(target_progress.get("evidence", []))
        evidence.append(EvidenceItem(
            messageIndex=message_count,
            role="user",
            label=eval_result["behavior_label"][:50] or "behavior",
            score=behavior_score,
        ))
        target_progress["evidence"] = evidence

    if success_score > 0:
        evidence = list(target_progress.get("evidence", []))
        evidence.append(EvidenceItem(
            messageIndex=message_count,
            role="ai",
            label=eval_result["success_label"][:50] or "success",
            score=success_score,
        ))
        target_progress["evidence"] = evidence

    # --- Achievement check ------------------------------------------------
    evidence = target_progress.get("evidence", [])
    evidence_count = len(evidence)
    strong_evidence_count = len([e for e in evidence if e.get("score", 0) >= strong_evidence_score])

    behavior_met = (
        behavior_score >= behavior_threshold
        and target_progress.get("confidence", 0) >= behavior_threshold
    )

    success_met = (
        success_score >= success_threshold
        if current_goal.get("successIndicators")
        else behavior_score >= behavior_threshold + 0.1
    )

    has_enough_evidence = (
        evidence_count >= min_evidence_count
        and strong_evidence_count >= min_strong_evidence_count
    )

    logger.info(
        f"[{session_id}] Goal #{goal_num} evaluation: "
        f"behavior={behavior_score:.2f}, success={success_score:.2f}, "
        f"behavior_met={behavior_met}, success_met={success_met}, "
        f"evidence={evidence_count}/{min_evidence_count}, "
        f"strong={strong_evidence_count}/{min_strong_evidence_count}"
    )
    if eval_result.get("reasoning"):
        logger.debug(f"[{session_id}] Eval reasoning: {eval_result['reasoning']}")

    if behavior_met and success_met and has_enough_evidence:
        if target_progress.get("status") != "achieved":
            target_progress["status"] = "achieved"
            target_progress["achievedAt"] = datetime.now().isoformat()
            logger.info(f"[{session_id}] Goal #{goal_num} ACHIEVED!")

    progress[progress_idx] = target_progress

    achieved_count = len([p for p in progress if p.get("status") == "achieved"])
    total_count = len(goals)

    trace = _add_trace(
        state, "evaluate_goals", start_time,
        f"goal #{goal_num}, behavior={behavior_score:.2f}, success={success_score:.2f}",
        f"status={target_progress.get('status')}, {achieved_count}/{total_count} achieved",
    )

    return {
        "goal_progress": progress,
        "evaluation_complete": True,
        "needs_evaluation": False,
        "node_trace": trace,
    }
